crosshair/stubs_parser.py

```diff
@@ -0,0 +1,289 @@
+from __future__ import annotations
+import ast
+from collections.abc import __all__ as abc_all
+from inspect import Parameter, Signature, signature
+from pathlib import Path
+import re
+import sys
+from typeshed_client import get_stub_ast, get_stub_file, get_search_context  # type: ignore
+from typing import (  # type: ignore
+    Any,
+    Callable,
+    Dict,
+    List,
+    Optional,
+    Tuple,
+    Union,
+    __all__ as typing_all,
+)
+
+from crosshair.util import debug
+
+
+def signature_from_stubs(fn: Callable) -> Optional[Signature]:
+    """
+    Try to find a signature for the given function in the stubs.
+
+    Note: this feature is only available for Python >= 3.8.
+
+    :param fn: The function to lookup a signature for.
+    :return: The signature found, if any.
+    """
+    # ast.get_source_segment requires Python 3.8
+    if sys.version_info < (3, 8):
+        return None
+    if not getattr(fn, "__module__", None) or not getattr(fn, "__qualname__", None):
+        return None
+    # Use the `qualname` to find the function inside its module.
+    path_in_module: List[str] = fn.__qualname__.split(".")
+    # Find the stub_file and corresponding AST using `typeshed_client`.
+    search_path = [Path(path) for path in sys.path if path]
+    search_context = get_search_context(search_path=search_path)
+    stub_file = get_stub_file(fn.__module__, search_context=search_context)
+    module = get_stub_ast(fn.__module__, search_context=search_context)
+    if not stub_file or not module or not isinstance(module, ast.Module):
+        debug("No stub found for module", fn.__module__)
+        return None
+    glo = globals().copy()
+    return _sig_from_ast(module.body, path_in_module, stub_file.read_text(), glo)
+
+
+def _sig_from_ast(
```

Comment:
There are similarities: For example `typeshed_client.get_stub_names("random")` returns the AST of all objects in the `random` module. This way one can easily find the `Random` object. You can even go one step further using `resolver = typeshed_client.Resolver()` and `resolver.get_fully_qualified_name("random.Random")`.
I first tried it this way, but the problem is that some parameter annotations might not be defined in the current namespace. So evaluating the function type annotations might fail. This is the reason why I go through the entire AST of the module to find the function and I execute the import statements and `TypeVar` definitions found on the way to have a namespace as close as possible as what is required.
I'm not fully satisfied of doing it this way, but I did not find a better way. I think that since stub files are not supposed to be used at runtime, there is no very clean way of getting the signature of a function.

Reply:
Ah, so you already tried it; that's good enough for me!